File: src/cpp/src/operations/curvature.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//

// System includes
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <limits>
#include <stdexcept>
#include <vector>

// Project includes
#include "meshioplusplus/operations/curvature.hpp"
#include "meshioplusplus/detail/data_ops.hpp"
#include "meshioplusplus/detail/geometry.hpp"
#include "meshioplusplus/detail/surface_distance.hpp"
#include "meshioplusplus/log.hpp"
#include "meshioplusplus/parallel.hpp"

namespace meshioplusplus {
namespace {

using detail::Vec3;

constexpr double kCurvTwoPi = 6.283185307179586476925286766559;
constexpr double kCurvPi = 3.141592653589793238462643383279;

/// One triangle corner's angle and the cotangent of it.
///
/// Both come from the same `(cross, dot)` pair, which is what keeps them
/// consistent. The angle is `atan2(|cross|, dot)` rather than `acos` of a
/// clamped ratio: `detail::sd_corner_angle` takes the `acos` route because a
/// pseudonormal weight does not care about the last few digits, but an angle
/// DEFECT is a sum of angles minus `2*pi`, so the digits are exactly what
/// survives -- do not unify the two.
///
/// The cotangent is `dot / |cross|`, trig-free and correctly NEGATIVE at an
/// obtuse corner. Never clamp it: clamping is the usual "fix" and it destroys
/// the operator's linear precision, which is the whole reason to use cotangent
/// weights rather than a uniform graph Laplacian.
struct CurvCorner {
    double mAngle = 0.0;
    double mCotangent = 0.0;
};

CurvCorner curv_corner(const Vec3& rApex, const Vec3& rB, const Vec3& rC) {
    const Vec3 u = detail::vec3_sub(rB, rApex);
    const Vec3 v = detail::vec3_sub(rC, rApex);
    const Vec3 cross = detail::vec3_cross(u, v);
    const double cross_norm = std::sqrt(detail::vec3_norm_sq(cross));
    const double dot = detail::vec3_dot(u, v);
    CurvCorner out;
    out.mAngle = std::atan2(cross_norm, dot);
    out.mCotangent = cross_norm > 0.0 ? dot / cross_norm : 0.0;
    return out;
}

/// Everything one serial pass over the triangles accumulates.
struct CurvAccumulators {
    std::vector<double> mAngleSum;  ///< sum of incident corner angles
    std::vector<Vec3> mLaplacian;   ///< the cotangent Laplacian of the positions
    std::vector<double> mArea;      ///< the dual area
    std::vector<Vec3> mNormal;      ///< area-weighted vertex normal
    std::vector<char> mTouched;     ///< referenced by at least one live triangle
    std::int64_t mNumDegenerate = 0;
};

/// The per-vertex dual-area contributions of one triangle.
///
/// Mixed Voronoi (Meyer et al. 2003): a non-obtuse triangle contributes
/// `(1/8) * (cot(alpha) * |e1|^2 + cot(beta) * |e2|^2)` to each of its corners,
/// where the cotangents are at the two corners OPPOSITE the respective edges.
/// An obtuse triangle has no valid Voronoi region, so it contributes half its
/// area at the obtuse corner and a quarter at each of the others.
void curv_dual_area(const Vec3& rA, const Vec3& rB, const Vec3& rC, const CurvCorner& rCa,
                    const CurvCorner& rCb, const CurvCorner& rCc, double TriArea,
                    CurvatureDualArea Mode, double* pOut) {
    if (Mode == CurvatureDualArea::Barycentric) {
        // Branch-free, hence bit-exactly twinnable in numpy.
        const double third = TriArea / 3.0;
        pOut[0] = third;
        pOut[1] = third;
        pOut[2] = third;
        return;
    }
    const bool obtuse_a = rCa.mAngle > 0.5 * kCurvPi;
    const bool obtuse_b = rCb.mAngle > 0.5 * kCurvPi;
    const bool obtuse_c = rCc.mAngle > 0.5 * kCurvPi;
    if (obtuse_a || obtuse_b || obtuse_c) {
        pOut[0] = obtuse_a ? 0.5 * TriArea : 0.25 * TriArea;
        pOut[1] = obtuse_b ? 0.5 * TriArea : 0.25 * TriArea;
        pOut[2] = obtuse_c ? 0.5 * TriArea : 0.25 * TriArea;
        return;
    }
    const double ab = detail::vec3_norm_sq(detail::vec3_sub(rB, rA));
    const double bc = detail::vec3_norm_sq(detail::vec3_sub(rC, rB));
    const double ca = detail::vec3_norm_sq(detail::vec3_sub(rA, rC));
    // Edge AB is opposite corner C, BC opposite A, CA opposite B.
    pOut[0] = 0.125 * (rCc.mCotangent * ab + rCb.mCotangent * ca);
    pOut[1] = 0.125 * (rCc.mCotangent * ab + rCa.mCotangent * bc);
    pOut[2] = 0.125 * (rCa.mCotangent * bc + rCb.mCotangent * ca);
}

/// The serial accumulation pass.
///
/// Serial deliberately: this is a scatter into shared per-vertex slots, so a
/// `parallel_for` would be both racy and -- once made safe -- order-dependent
/// in the last bits. The finalize pass below is the parallel half, and it is
/// per-vertex and order-independent. The same phase split `surface_distance`
/// and `surface.cpp` already use.
CurvAccumulators curv_accumulate(const detail::TriangleSoup& rSoup, CurvatureDualArea Mode) {
    const std::size_t npts = rSoup.mPoints.size();
    CurvAccumulators acc;
    acc.mAngleSum.assign(npts, 0.0);
    acc.mLaplacian.assign(npts, Vec3{0.0, 0.0, 0.0});
    acc.mArea.assign(npts, 0.0);
    acc.mNormal.assign(npts, Vec3{0.0, 0.0, 0.0});
    acc.mTouched.assign(npts, 0);

    const std::size_t ntri = rSoup.NumTriangles();
    for (std::size_t t = 0; t < ntri; ++t) {
        const std::array<std::int64_t, 3>& v = rSoup.mVertices[t];
        const Vec3& a = rSoup.mCorners[t * 3 + 0];
        const Vec3& b = rSoup.mCorners[t * 3 + 1];
        const Vec3& c = rSoup.mCorners[t * 3 + 2];
        const Vec3 cross = detail::vec3_cross(detail::vec3_sub(b, a), detail::vec3_sub(c, a));
        // Exactly `soup_quality`'s predicate, so the two cannot disagree about
        // which triangles are degenerate.
        if (!(detail::vec3_norm_sq(cross) > 0.0)) {
            ++acc.mNumDegenerate;
            continue;
        }
        const double tri_area = 0.5 * std::sqrt(detail::vec3_norm_sq(cross));

        const CurvCorner ca = curv_corner(a, b, c);
        const CurvCorner cb = curv_corner(b, c, a);
        const CurvCorner cc = curv_corner(c, a, b);

        double dual[3] = {0.0, 0.0, 0.0};
        curv_dual_area(a, b, c, ca, cb, cc, tri_area, Mode, dual);

        const Vec3* corner[3] = {&a, &b, &c};
        const CurvCorner* angle[3] = {&ca, &cb, &cc};
        for (std::size_t i = 0; i < 3; ++i) {
            const std::size_t vi = static_cast<std::size_t>(v[i]);
            acc.mTouched[vi] = 1;
            acc.mAngleSum[vi] += angle[i]->mAngle;
            acc.mArea[vi] += dual[i];
            acc.mNormal[vi] = detail::vec3_add(acc.mNormal[vi], cross);  // area-weighted
        }
        // The cotangent Laplacian of the positions. Edge (i, j)'s weight is the
        // cotangent at the corner OPPOSITE it; each triangle contributes one
        // half of each of its three edges' weights, and the two triangles
        // sharing an edge complete the usual (cot a + cot b) / 2.
        for (std::size_t e = 0; e < 3; ++e) {
            const std::size_t i = e;
            const std::size_t j = (e + 1) % 3;
            const std::size_t k = (e + 2) % 3;  // the opposite corner
            const double w = 0.5 * angle[k]->mCotangent;
            const Vec3 d = detail::vec3_sub(*corner[i], *corner[j]);
            const std::size_t vi = static_cast<std::size_t>(v[i]);
            const std::size_t vj = static_cast<std::size_t>(v[j]);
            acc.mLaplacian[vi] = detail::vec3_add(acc.mLaplacian[vi], detail::vec3_scale(d, w));
            acc.mLaplacian[vj] = detail::vec3_add(acc.mLaplacian[vj], detail::vec3_scale(d, -w));
        }
    }
    return acc;
}

/// Which vertices sit on a boundary, from the SAME edge map `soup_quality` and
/// `repair` read -- so the three cannot disagree about what a boundary is.
std::vector<char> curv_boundary_vertices(const detail::TriangleSoup& rSoup,
                                         const detail::SurfaceEdgeMap& rEdges) {
    std::vector<char> is_boundary(rSoup.mPoints.size(), 0);
    for (const auto& kv : rEdges) {
        if (kv.second.mUsed != 1)
            continue;
        // A self-loop can only come from a triangle with a repeated corner,
        // which is degenerate and was skipped by the accumulation pass. Letting
        // it mark a boundary would NaN out a perfectly good vertex because of a
        // triangle that contributed nothing -- and `soup_quality` counts it as
        // a boundary edge (correctly, for its own purpose), so the shared map
        // carries it and this is where it has to be filtered.
        if (kv.first[0] == kv.first[1])
            continue;
        is_boundary[static_cast<std::size_t>(kv.first[0])] = 1;
        is_boundary[static_cast<std::size_t>(kv.first[1])] = 1;
    }
    return is_boundary;
}

}  // namespace

CurvatureDualArea curvature_dual_area_from_name(const std::string& rName) {
    if (rName == "mixed-voronoi")
        return CurvatureDualArea::MixedVoronoi;
    if (rName == "barycentric")
        return CurvatureDualArea::Barycentric;
    throw std::invalid_argument("meshio++: curvature: unknown dual area '" + rName +
                                "' (expected 'mixed-voronoi' or 'barycentric')");
}

const char* curvature_dual_area_name(CurvatureDualArea Mode) {
    switch (Mode) {
        case CurvatureDualArea::MixedVoronoi:
            return "mixed-voronoi";
        case CurvatureDualArea::Barycentric:
            return "barycentric";
    }
    return "mixed-voronoi";
}

CurvatureResult compute_curvature(const Mesh& rMesh, const CurvatureOptions& rOptions) {
    const detail::TriangleSoup soup = detail::build_triangle_soup(rMesh, rOptions.mRegion);
    const detail::SurfaceEdgeMap edges = detail::build_surface_edges(soup);

    CurvatureResult out;
    out.mQuality = detail::soup_quality(soup);
    out.mMesh = detail::clone_mesh(
        rMesh, [](DataLocation, const std::string&, std::string&) { return true; });

    const std::size_t npts = soup.mPoints.size();
    const CurvAccumulators acc = curv_accumulate(soup, rOptions.mDualArea);
    const std::vector<char> is_boundary = curv_boundary_vertices(soup, edges);
    out.mNumDegenerate = acc.mNumDegenerate;

    const double nan = std::numeric_limits<double>::quiet_NaN();
    NDArray mean(DType::Float64, {npts});
    NDArray gauss(DType::Float64, {npts});
    NDArray area(DType::Float64, {npts});
    NDArray principal(DType::Float64, {npts, 2});
    double* p_mean = mean.As<double>();
    double* p_gauss = gauss.As<double>();
    double* p_area = area.As<double>();
    double* p_principal = principal.As<double>();

    // Per-vertex and order-independent: the parallel half of the phase split.
    parallel_for(npts, [&](std::size_t i) {
        const double a = acc.mArea[i];
        p_area[i] = acc.mTouched[i] ? a : nan;
        if (!acc.mTouched[i] || !(a > 0.0) || (is_boundary[i] && !rOptions.mIncludeBoundary)) {
            p_mean[i] = nan;
            p_gauss[i] = nan;
            p_principal[i * 2] = nan;
            p_principal[i * 2 + 1] = nan;
            return;
        }
        // The geodesic form on a boundary vertex, the closed one otherwise.
        const double turn = is_boundary[i] ? kCurvPi : kCurvTwoPi;
        const double k = (turn - acc.mAngleSum[i]) / a;
        // ||L p|| / 2A is |H|; the sign comes from whether the mean-curvature
        // normal points along the surface normal or against it. That is what
        // makes H orientation-dependent and K not.
        const Vec3& lp = acc.mLaplacian[i];
        const double h_mag = 0.5 * std::sqrt(detail::vec3_norm_sq(lp)) / a;
        // `mLaplacian` accumulates sum(w * (p_i - p_j)), which is the NEGATIVE
        // of the usual Laplacian sum(w * (p_j - p_i)), so the mean-curvature
        // normal is -lp. On a convex outward-oriented surface that points along
        // the vertex normal, and H must come out positive there.
        const double sign = detail::vec3_dot(lp, acc.mNormal[i]) > 0.0 ? 1.0 : -1.0;
        const double h = h_mag * sign;
        p_mean[i] = h;
        p_gauss[i] = k;
        // k1, k2 = H +- sqrt(H^2 - K). The radicand is negative only through
        // discretization error, so clamp rather than produce NaN.
        const double disc = h * h - k;
        const double root = std::sqrt(disc > 0.0 ? disc : 0.0);
        p_principal[i * 2] = h + root;
        p_principal[i * 2 + 1] = h - root;
    });

    for (std::size_t i = 0; i < npts; ++i) {
        if (!acc.mTouched[i])
            ++out.mNumIsolated;
        else if (is_boundary[i])
            ++out.mNumBoundary;
        out.mTotalAngleDefect += acc.mTouched[i] ? (kCurvTwoPi - acc.mAngleSum[i]) : 0.0;
    }

    if (rOptions.mMean)
        out.mMesh.AddPointData(kCurvatureMeanName, std::move(mean));
    if (rOptions.mGaussian)
        out.mMesh.AddPointData(kCurvatureGaussianName, std::move(gauss));
    if (rOptions.mRecordArea)
        out.mMesh.AddPointData(kCurvatureAreaName, std::move(area));
    if (rOptions.mRecordPrincipal)
        out.mMesh.AddPointData(kCurvaturePrincipalName, std::move(principal));

    if (out.mQuality.mInconsistentPairs != 0)
        log::warn(
            "curvature: {} edge pair(s) wind the same way, so the sign of "
            "'{}' is not trustworthy; run repair(mesh, {{.mFixOrientation = true}}) first",
            out.mQuality.mInconsistentPairs, kCurvatureMeanName);
    return out;
}

}  // namespace meshioplusplus

File: src/python/meshioplusplus/_curvature.py
"""Per-vertex mean and Gaussian curvature of a surface, by the standard
discrete estimators: the angle defect for ``K`` and the cotangent
Laplace-Beltrami operator for ``H``.

These are the estimators the discrete-differential-geometry literature's
convergence results are about, and the ones NVIDIA PhysicsNeMo's own
``gaussian_curvature_vertices``/``mean_curvature_vertices`` use, so the numbers
here are comparable with a model's. Curvature is the signed distance's natural
companion as a node feature: :func:`meshioplusplus.sample_distance` says how
far a point is from the surface, this says how the surface bends there.

**Deliberately NOT** :func:`meshioplusplus.remesh`'s estimator, which fits an
osculating paraboloid over each 1-ring to drive its own gradation and
anisotropic metric. That one returns only the larger-magnitude principal
curvature as a *magnitude*, runs on remesh's subdivided working copy rather
than the caller's mesh, and has no structural invariant to test against. It is
left exactly as it is; this is a second, independent estimator.

**The oracle is Gauss-Bonnet.** For a closed surface the angle defects sum to
``2*pi*chi`` -- ``4*pi`` for a sphere -- whatever the tessellation and
whichever dual area. The result carries that sum as ``total_angle_defect``
precisely so the invariant is checkable from every binding and not only from a
gtest.

**``H`` is orientation-dependent and ``K`` is not.** The mean curvature's sign
comes from the surface's own winding, so a mesh whose facets disagree about
which side is out yields sign-flipped patches with no error raised. That is why
the report carries the input's ``quality``: check ``quality["inconsistent_pairs"]``
before trusting a sign. This operation never silently repairs its input.

### Where byte-parity stops

``dual_area="barycentric"`` is branch-free -- every vertex gets a third of each
incident triangle's area -- so this module implements it as a genuine numpy
twin of the C++ core and ``tests/python/test_curvature.py`` pins the two
against each other to a tight tolerance (not exact equality: the one
``atan2`` in the angle-defect computation goes through ``std::atan2`` on the
C++ side and numpy's vectorized ``arctan2`` on the Python side, and those two
transcendental-function implementations carry no cross-library bit-exactness
guarantee -- measured at up to ~1e-14 relative difference on some platforms).

``dual_area="mixed-voronoi"`` (the default) is not twinned and **raises**
:class:`NotImplementedError`: its obtuse/non-obtuse test is a discrete branch
on a sign, so two implementations could land on opposite sides for a
right-angled triangle and then disagree macroscopically rather than in the last
ulp -- exactly the case ``_smooth.py`` refuses for its inversion guard and
``_sdf.py`` for ``sign="winding-number"``. The raise only fires when the
compiled core is genuinely unavailable.

Public API:

* :func:`compute_curvature` -- the estimator.
"""

from __future__ import annotations

import numpy as np

from ._regions import block_bases

_PREFIX = "meshio++: curvature: "

#: Point data written by :func:`compute_curvature`. Twins of ``curvature.hpp``'s
#: ``kCurvature*Name`` constants -- keep the two in step.
MEAN_NAME = "curvature:mean"
GAUSSIAN_NAME = "curvature:gaussian"
AREA_NAME = "curvature:area"
PRINCIPAL_NAME = "curvature:principal"

_DUAL_AREAS = ("mixed-voronoi", "barycentric")

_TWO_PI = 6.283185307179586476925286766559
_PI = 3.141592653589793238462643383279


def _validate(dual_area):
    if dual_area not in _DUAL_AREAS:
        raise ValueError(
            f"{_PREFIX}unknown dual area '{dual_area}' "
            "(expected 'mixed-voronoi' or 'barycentric')"
        )


def _region_mask(mesh, region):
    """Global block-major cell mask for a named ``Cell`` region.

    The numpy twin of ``sd_region_mask``: same lookup, same message shape, and
    the same "out-of-range entries are ignored" rule.
    """
    if not region:
        return None
    for r in getattr(mesh, "regions", ()) or ():
        if r.name == region and r.kind == "cell":
            bases = block_bases(mesh.cells)
            total = int(bases[-1]) + (
                len(mesh.cells[-1].data) if len(mesh.cells) else 0
            )
            mask = np.zeros(total, dtype=bool)
            e = np.asarray(r.entries, dtype=np.int64).ravel()
            e = e[(e >= 0) & (e < total)]
            mask[e] = True
            return mask
    have = ", ".join(sorted({r.name for r in (getattr(mesh, "regions", ()) or ())}))
    raise ValueError(
        f"{_PREFIX}no cell region named '{region}' "
        f"(available: {have if have else 'none'})"
    )


def _dot(u, v):
    """``vec3_dot``'s own left-to-right order, never ``np.sum``/``einsum``.

    ``detail/geometry.hpp`` spells it ``a[0]*b[0] + a[1]*b[1] + a[2]*b[2]``;
    numpy's reductions are free to pair-sum, which differs in the last bit.
    """
    return u[:, 0] * v[:, 0] + u[:, 1] * v[:, 1] + u[:, 2] * v[:, 2]


def _cross(u, v):
    """``vec3_cross``'s own component order."""
    return np.column_stack(
        [
            u[:, 1] * v[:, 2] - u[:, 2] * v[:, 1],
            u[:, 2] * v[:, 0] - u[:, 0] * v[:, 2],
            u[:, 0] * v[:, 1] - u[:, 1] * v[:, 0],
        ]
    )


def _corner(apex, b, c):
    """``curv_corner``'s angle and cotangent, from one ``(cross, dot)`` pair.

    The angle is ``atan2(|cross|, dot)`` and deliberately **not** ``acos`` of a
    clamped ratio: an angle defect is a sum of angles minus ``2*pi``, so the
    digits are exactly what survives. The cotangent is ``dot / |cross|``,
    trig-free, correctly negative at an obtuse corner, and **never clamped** --
    clamping destroys the operator's linear precision, which is the whole
    reason for cotangent weights over a uniform graph Laplacian.
    """
    u = b - apex
    v = c - apex
    cr = _cross(u, v)
    cross_norm = np.sqrt(_dot(cr, cr))
    dot = _dot(u, v)
    angle = np.arctan2(cross_norm, dot)
    cot = np.where(
        cross_norm > 0.0, dot / np.where(cross_norm > 0.0, cross_norm, 1.0), 0.0
    )
    return angle, cot


def _soup(mesh, region):
    """The triangle soup ``build_triangle_soup`` builds, region filter included.

    Reuses ``_sdf._soup`` rather than transcribing the fan a second time, so
    the two cannot disagree about which diagonal a quad is split on. The region
    filter is applied to the *triangles* by their source cell, which is
    equivalent to filtering cells before fanning because the fan is per-cell.
    """
    from ._sdf import _soup as sdf_soup

    points, verts, corners, source = sdf_soup(mesh)
    mask = _region_mask(mesh, region)
    if mask is not None and len(verts):
        keep = mask[source]
        verts = verts[keep]
        corners = corners[keep]
        source = source[keep]
    return points, verts, corners


def _boundary_vertices(npts, verts):
    """Vertices on an edge used by exactly one triangle.

    From the same edge set ``build_surface_edges`` builds -- **every** triangle,
    degenerate ones included, so the two cannot disagree about what a boundary
    is. Self-loops are filtered out for the reason ``curv_boundary_vertices``
    states: they come only from a repeated corner, whose triangle contributed
    nothing, and letting one mark a boundary would NaN out a good vertex.
    """
    out = np.zeros(npts, dtype=bool)
    if not len(verts):
        return out
    lo = np.minimum(verts, np.roll(verts, -1, axis=1)).ravel()
    hi = np.maximum(verts, np.roll(verts, -1, axis=1)).ravel()
    keys = lo * (np.int64(npts) + 1) + hi
    uniq, counts = np.unique(keys, return_counts=True)
    once = uniq[counts == 1]
    if not len(once):
        return out
    u_lo = once // (np.int64(npts) + 1)
    u_hi = once % (np.int64(npts) + 1)
    real = u_lo != u_hi
    out[u_lo[real]] = True
    out[u_hi[real]] = True
    return out


def _curvature_py(
    mesh,
    mean,
    gaussian,
    dual_area,
    include_boundary,
    record_area,
    record_principal,
    region,
):
    """The numpy twin. ``barycentric`` only -- see the module docstring."""
    if dual_area != "barycentric":
        raise NotImplementedError(
            f"{_PREFIX}the numpy reference does not implement "
            f"dual_area='{dual_area}': its obtuse/non-obtuse test is a discrete "
            "branch on a sign, so two implementations could land on opposite "
            "sides for a right-angled triangle and then disagree "
            "macroscopically rather than in the last ulp. Use the compiled "
            "core for it, or dual_area='barycentric', which is branch-free and "
            "bit-exactly twinned."
        )

    points, verts, corners = _soup(mesh, region)
    npts = len(points)
    nan = float("nan")

    angle_sum = np.zeros(npts, dtype=np.float64)
    lap = np.zeros((npts, 3), dtype=np.float64)
    area = np.zeros(npts, dtype=np.float64)
    normal = np.zeros((npts, 3), dtype=np.float64)
    touched = np.zeros(npts, dtype=bool)
    num_degenerate = 0

    is_boundary = _boundary_vertices(npts, verts)

    if len(verts):
        a, b, c = corners[:, 0], corners[:, 1], corners[:, 2]
        cr = _cross(b - a, c - a)
        cr_sq = _dot(cr, cr)
        # Exactly `soup_quality`'s predicate, so the two cannot disagree about
        # which triangles are degenerate. Written as `not (x > 0)` so a NaN
        # counts as degenerate, matching the C++ `!(... > 0.0)`.
        alive = cr_sq > 0.0
        num_degenerate = int(np.count_nonzero(~alive))

        a, b, c = a[alive], b[alive], c[alive]
        cr, cr_sq = cr[alive], cr_sq[alive]
        v = verts[alive]
        tri_area = 0.5 * np.sqrt(cr_sq)

        ang_a, cot_a = _corner(a, b, c)
        ang_b, cot_b = _corner(b, c, a)
        ang_c, cot_c = _corner(c, a, b)

        # Branch-free barycentric dual area: a third of the triangle, each.
        third = tri_area / 3.0

        # `mAngleSum`/`mArea`/`mNormal` take their contributions in (triangle,
        # corner) order. `np.add.at` is unbuffered and applies them in
        # index-array order, which is what reproduces the C++ serial scatter
        # bit for bit -- never `np.bincount`/`reduceat`, which reassociate.
        idx = v.ravel()
        touched[idx] = True
        np.add.at(angle_sum, idx, np.column_stack([ang_a, ang_b, ang_c]).ravel())
        np.add.at(area, idx, np.repeat(third, 3))
        np.add.at(normal, idx, np.repeat(cr, 3, axis=0))

        # The cotangent Laplacian takes its contributions in a DIFFERENT order:
        # per triangle e = 0, 1, 2 with a +/- pair each, so the index sequence
        # is v0, v1, v1, v2, v2, v0. Edge (i, j)'s weight is the cotangent at
        # the corner opposite it. Build the interleaving explicitly; do not
        # "simplify" it to the (triangle, corner) order above.
        w0 = 0.5 * cot_c
        w1 = 0.5 * cot_a
        w2 = 0.5 * cot_b
        d0, d1, d2 = a - b, b - c, c - a
        lidx = np.column_stack(
            [v[:, 0], v[:, 1], v[:, 1], v[:, 2], v[:, 2], v[:, 0]]
        ).ravel()
        lval = np.empty((len(v), 6, 3), dtype=np.float64)
        lval[:, 0] = d0 * w0[:, None]
        lval[:, 1] = d0 * (-w0)[:, None]
        lval[:, 2] = d1 * w1[:, None]
        lval[:, 3] = d1 * (-w1)[:, None]
        lval[:, 4] = d2 * w2[:, None]
        lval[:, 5] = d2 * (-w2)[:, None]
        np.add.at(lap, lidx, lval.reshape(-1, 3))

    # The parallel, per-vertex, order-independent half of the phase split.
    out_area = np.where(touched, area, nan)
    turn = np.where(is_boundary, _PI, _TWO_PI)
    bad = (~touched) | ~(area > 0.0) | (is_boundary & (not include_boundary))
    safe = np.where(area > 0.0, area, 1.0)

    k = (turn - angle_sum) / safe
    h_mag = 0.5 * np.sqrt(_dot(lap, lap)) / safe
    # `lap` accumulates sum(w * (p_i - p_j)), the NEGATIVE of the usual
    # Laplacian, so the mean-curvature normal is -lap; on a convex
    # outward-oriented surface that points along the vertex normal and H must
    # come out positive there.
    sign = np.where(_dot(lap, normal) > 0.0, 1.0, -1.0)
    h = h_mag * sign
    # k1, k2 = H +- sqrt(H^2 - K). The radicand is negative only through
    # discretization error, so clamp rather than produce NaN.
    disc = h * h - k
    root = np.sqrt(np.where(disc > 0.0, disc, 0.0))

    k = np.where(bad, nan, k)
    h = np.where(bad, nan, h)
    principal = np.column_stack(
        [np.where(bad, nan, h + root), np.where(bad, nan, h - root)]
    )

    num_isolated = int(np.count_nonzero(~touched))
    num_boundary = int(np.count_nonzero(touched & is_boundary))
    # A plain left-to-right serial sum, as the C++ finalize loop does.
    # `np.sum` pair-sums and would differ; `cumsum` is sequential by
    # construction, since it must produce every prefix.
    contrib = np.where(touched, _TWO_PI - angle_sum, 0.0)
    total = float(np.cumsum(contrib)[-1]) if npts else 0.0

    out = mesh.copy() if hasattr(mesh, "copy") else mesh
    if mean:
        out.point_data[MEAN_NAME] = h
    if gaussian:
        out.point_data[GAUSSIAN_NAME] = k
    if record_area:
        out.point_data[AREA_NAME] = out_area
    if record_principal:
        out.point_data[PRINCIPAL_NAME] = principal

    from ._sdf import _watertight_py

    return (
        out,
        {
            "num_boundary": num_boundary,
            "num_isolated": num_isolated,
            "num_degenerate": num_degenerate,
            "total_angle_defect": total,
            "quality": _watertight_py(mesh),
        },
    )


def compute_curvature(
    mesh,
    mean=True,
    gaussian=True,
    dual_area="mixed-voronoi",
    include_boundary=False,
    record_area=False,
    record_principal=False,
    region="",
    return_report=False,
):
    """Per-vertex mean and Gaussian curvature of a surface mesh.

    Triangles come from the same fan :func:`meshioplusplus.convert_cells`'s
    ``simplexify`` uses, so a quad mesh's curvature is the curvature of its
    canonical triangulation. A volume or polyhedron block is refused by name
    pointing at ``extract_surface``, a higher-order one pointing at
    ``linearize``.

    :param mesh: a surface mesh (never modified).
    :param mean: attach ``curvature:mean``.
    :param gaussian: attach ``curvature:gaussian``.
    :param dual_area: ``"mixed-voronoi"`` (default) or ``"barycentric"``.
    :param include_boundary: compute a (biased) value at boundary vertices
        instead of leaving them NaN. Isolated vertices are NaN either way.
    :param record_area: also attach ``curvature:area``, the dual area each
        curvature was divided by.
    :param record_principal: also attach ``curvature:principal``, the ``(n, 2)``
        pair ``k1 >= k2`` recovered as ``H +- sqrt(H^2 - K)``.
    :param region: restrict to this named ``Cell`` region; ``""`` takes every
        surface cell.
    :param return_report: also return the counters and the input's surface
        ``quality``.
    :returns: the mesh carrying the requested ``point_data``, or
        ``(mesh, report)`` when ``return_report`` is set.
    :raises ValueError: on a non-surface input (naming the fix) or an unknown
        region name.
    :raises NotImplementedError: when the compiled core is unavailable and
        ``dual_area`` is not ``"barycentric"`` -- see this module's docstring.

    For a closed surface ``report["total_angle_defect"]`` is ``2*pi*chi``
    exactly, whatever the tessellation and whichever dual area: the
    Gauss-Bonnet invariant, and the cheapest check that a result is sane.
    """
    _validate(dual_area)

    out = None
    report = None
    try:
        from . import _core

        res = _core.compute_curvature(
            mesh,
            bool(mean),
            bool(gaussian),
            dual_area,
            bool(include_boundary),
            bool(record_area),
            bool(record_principal),
            str(region),
        )
        out = res["mesh"]
        report = {
            "num_boundary": res["num_boundary"],
            "num_isolated": res["num_isolated"],
            "num_degenerate": res["num_degenerate"],
            "total_angle_defect": res["total_angle_defect"],
            "quality": res["quality"],
        }
    except (ValueError, TypeError):
        # A genuine user error must not fall through to the numpy path, which
        # would either raise something less helpful or silently differ.
        raise
    except Exception:
        out = None

    if out is None:
        out, report = _curvature_py(
            mesh,
            bool(mean),
            bool(gaussian),
            dual_area,
            bool(include_boundary),
            bool(record_area),
            bool(record_principal),
            str(region),
        )

    if report["quality"]["inconsistent_pairs"]:
        from ._common import warn

        warn(
            f"curvature: {report['quality']['inconsistent_pairs']} edge pair(s) "
            f"wind the same way, so the sign of '{MEAN_NAME}' is not "
            "trustworthy; run repair(mesh, fix_orientation=True) first"
        )

    return (out, report) if return_report else out


__all__ = ["compute_curvature"]
